critiquebrainz/db/users.py

```diff
@@ -70,3 +73,543 @@ def get_many_by_mb_username(usernames):
+
+    Args:
+        limit(int, optional): Number of users to be returned
+        offset(int, optional): Number of initial users to skip
+
+    Returns:
+        List of dictionaries of users with the following structure:
+        {
+            "id": (uuid)
+            "display_name": (str)
+            "email": (str)
+            "created": (datetime)
+            "musicbrainz_id": (str)
+            "show_gravatar": (bool)
+            "is_blocked": (bool)
+        }
+    """
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT id, display_name, email, created,
+                   musicbrainz_id, show_gravatar, is_blocked
+              FROM "user"
+             LIMIT :limit
+            OFFSET :offset
+        """), {
+            "limit": limit,
+            "offset": offset
+        })
+        rows = result.fetchall()
+        rows = [dict(row) for row in rows]
+    return rows
+
+
+def unblock(user_id):
+    """Unblock user (admin only)
+
+    Args:
+        user_id(uuid): ID of user to be unblocked
+    """
+    with db.engine.connect() as connection:
+        connection.execute(sqlalchemy.text("""
+            UPDATE "user"
+               SET is_blocked = 'false'
+             WHERE id = :user_id
+        """), {
+            "user_id": user_id
+        })
+
+
+def block(user_id):
+    """Block user (admin only)
+
+    Args:
+        user_id(uuid): ID of user to be blocked
+    """
+    with db.engine.connect() as connection:
+        connection.execute(sqlalchemy.text("""
+            UPDATE "user"
+               SET is_blocked = 'true'
+            WHERE id = :user_id
+        """), {
+            "user_id": user_id
+        })
+
+
+def has_voted(user_id, review_id):
+    """Check if a user has already voted a review
+
+    Args:
+        user_id(uuid): ID of the user
+        review_id(uuid): ID of the review
+
+    Returns:
+        (bool): True if has voted else False
+    """
+    last_revision = db_revision.get(review_id, limit=1)[0]
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT count(*)
+              FROM vote
+             WHERE revision_id = :revision_id
+               AND user_id = :user_id
+            """), {
+                "revision_id": last_revision['id'],
+                "user_id": user_id
+            })
+        count = result.fetchone()[0]
+    return count > 0
+
+
+def karma(user_id):
+    """Get the karma of a user
+
+    Args:
+        user_id(uuid): ID of the user
+
+    Returns:
+        karma(int): the karma of the user
+    """
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT review.user_id, vote
+              FROM ((vote
+         LEFT JOIN revision
+                ON revision.id = revision_id)
+         LEFT JOIN review
+                ON review.id = review_id)
```

Comment:
You might need to add another level of indentation by parentheses to make this easier to read.